### PR TITLE
Fix nested global blocks

### LINKED ISSUE #7409: [BC Break] Sulu 2.6: Possible bug and/or documentation needed? How to migrate to new global blocks?
| Q | A
| --- | ---
| Sulu Version | 2.6.1 (since 2.6.0)
| PHP Version | 8.2 (several versions)
| DB Version | MariaDB 10.3.39
| Browser Version | Chrome 124.0.6367.91

# Actual Behavior

After upgrading to Sulu 2.6 my included block types lead to different errors (on cache clear or in browser-console):

I have nested blocks that are all on their own xml-file. Like `blocks/block_text.xml`. Those are `type` elements with a `name` attribute like `name="block_text"` and they are included within the `types` node of other templates via `<xi:include href="blocks/block_text.xml"/>`. That worked well until Sulu 2.5:

```` xml
<?xml version="1.0" ?>
<type xmlns="http://schemas.sulu.io/template/template"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://schemas.sulu.io/template/template http://schemas.sulu.io/template/template-1.0.xsd"
      name="block_text">
  <meta>
    <title lang="en">Textblock</title>
  </meta>
  <properties>
    <property name="text" type="text_area">
      <meta>
        <title lang="en">Content</title>
      </meta>
    </property>
  </properties>
</type>
````

```` xml
<!-- … -->
  <types>
    <xi:include href="blocks/block_text.xml"/>
  </types>
<!-- … -->
````

After upgrading to Sulu 2.6, I get an error on cache clear with that:

>  [ERROR 1845] Element '{http://schemas.sulu.io/template/template}type': No matching global declaration available for the validation root. (in D:\Projects\sulu.dev.lan/config/templates/blocks\block_text.xml - line 5, column 0)

Interestingly, that error does not raise for elements (instead of blocks) that are defined and included the exact same way and it also does not raise, when I copy the type definition directly into its parent instead of including it via XInclude. **This is my workaround for now**.

Since the blocks have changed in Sulu 2.6 to global blocks, I then changed their definition according to the current documentation in https://docs.sulu.io/en/2.6/book/templates.html#using-global-blocks to the following code:

```` xml
<?xml version="1.0" ?>
<template xmlns="http://schemas.sulu.io/template/template"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://schemas.sulu.io/template/template http://schemas.sulu.io/template/template-1.0.xsd">
  <key>block_text</key>
  <meta>
    <title lang="en">Textblock</title>
  </meta>
  <properties>
    <property name="text" type="text_area">
      <meta>
        <title lang="en">Content</title>
      </meta>
    </property>
  </properties>
</template>
````

```` xml
<!-- … -->
  <types>
    <type ref="block_text"/>
  </types>
<!-- … -->
````

That removes the error on cache clear, but now, I cannot edit pages containing that block type any more. Instead I get the following error inside the browser console:

> uncaught (in promise) Error: can't resolve reference #/definitions/block_text from id #

Adding an name-attribute to the ´type´ node with the `ref` leads to an xml error, as that `ref` and `name` are not allowed together.

# Expected Behavior

I expected either the old `xi:include` way to still work for included block types or the new global block handling to work with my existing content. Or at least a documented way to migrate my existing content to work with the new global blocks.

# Steps to Reproduce

Have a Sulu 2.5 instance with nested block based content using xi:include and upgrade to Sulu 2.6. We have two problems here: The first is the one with the new `xi:include` error for included block types and the other one that the needed migration for existing content that is not documented.

# Possible Solutions

I don't know how to solve that, but I guess that the existing content has to be migrated to be compatible to the new global block handling. Some documentation on how to do that would be helpful. Better would be to change the implementation in a way that no migration would be needed at all when the blocks keep their name.

For now, my workaround is to not include my block types via `xi:include`, but to copy their definitions directly into the templates using them.

### LINKED ISSUE #7414: A global block in a global block leads to 500 errors
| Q | A
| --- | ---
| Sulu Version | 2.6.0
| PHP Version | 8.3

# Actual Behavior

If a global block (e.g. `foo`) is loaded into another global block (e.g. `bar`), this results in a 500 error for `/admin/config` due to a recursion. This was not a problem with includes.

# Steps to Reproduce

I have created a small example (with text_block and text_block**2**):
https://github.com/flogado/sulu-global-blocks/commit/adaab4a398e975de089e9265a019f0f1f74ff26c

# Possible Solutions

The recursion seems to happen in the `GlobalBlocksTypedFormMetadataVisitor::getGlobalBlockMetadata`.

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #7414 , fixes #7409
| Related issues/PRs | #7229 
| License | MIT
| Documentation PR | none

#### What's in this PR?

This PR fixed following issues:

* #7414
* #7409

#### Why?

This issues where introduced by #7229

### CODE REVIEW

**Comment by mamazu** [comment]
Now when I expand a global block type the list of properties is empty.

**Comment by wachterjohannes** [comment]
@mamazu seems to work for me pretty well:

image

Have you rebuild the admin dist files?

**Comment by wachterjohannes** [comment]
@mamazu see comment in the issue https://github.com/sulu/sulu/issues/7409#issuecomment-2109840871

**Comment by mamazu** [comment]
Yes. Oddly there are also no console errors this time when I expand a block which is empty. So no clue what's happening.

**Comment by wachterjohannes** [comment]
@mamazu does it work now?

**Comment by spackmat** [comment]
@wachterjohannes I also get no console errors now, bus still does not work.

**Comment by wachterjohannes** [comment]
@spackmat only to be clear - after updating the vendor to my branch - you have to remove the node_modules folder and run `npm install && npm run build` to be sure that my new files are used

**Comment by mamazu** [comment]
@wachterjohannes Still no change. Even ran the Sulu admin build update command.

**Comment by ghost** [comment]
Works for me, but there is an error with local and global blocks combined:
https://github.com/sulu/sulu/issues/7414#issuecomment-2109979442

**Comment by spackmat** [comment]
@wachterjohannes did it all again for you, still the same: When the blocks are within a section, there are no fields displayed. Outside of a section it works well for normal blocks, nested blocks do not work at all, showing the wrong name.

Looks like this (opened), when within a section:
image

And should look like this (when outside of a section:
image

And looks like this (without section) for nested blocks:
image
(outside of the nested block, the block is correctly named "Studienverlaufsplan", when not nested)

**Comment by wachterjohannes** [comment]
@spackmat me and @mamazu found the issue and pushed it to the branch - can you recheck?

**Comment by spackmat** [comment]
@wachterjohannes @mamazu Hey, the non nested blocks within sections do work now 👏

The nested blocks still fail the same way as before.

**Comment by spackmat** [comment]
@wachterjohannes with your last commit, the nested blocks also work. Nice! 😎

**Comment by wachterjohannes** [comment]
@spackmat thanks for the testing :)

@mamazu will you give it a try again?

**Comment by FlorianKoerner** [comment]
@wachterjohannes Looks good - even with nested blocks. Thank you! :)

File: src/Sulu/Bundle/AdminBundle/Tests/Functional/Controller/AdminControllerTest.php [hunk 1]
```diff
@@ -167,4 +167,48 @@ public function testGetMetaDataKeysOnly(): void
...
+        $response = $this->client->getResponse();
+        $this->assertHttpStatusCode(200, $response);
+        $json = $response->getContent();
+        $this->assertIsString($json);
+
+        $metaData = \json_decode($json, true, 512, \JSON_THROW_ON_ERROR);
+
+        $defaultMetadata = $metaData['types']['default'];
+        $defaultSchema = $defaultMetadata['schema'];
+        $this->assertArrayHasKey('text_block', $defaultSchema['definitions']);
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/AdminBundle/Tests/Functional/Controller/AdminControllerTest.php]: Why not do:

```
$this->assertSame($expectedSchema, $defaultMetadata['schema']);
```

Would show us unexpected changes in the schema and you don't have the new the phpstan issues.

File: phpstan-baseline.neon [hunk 2]
```diff
@@ -1455,6 +1455,81 @@ parameters:
 			count: 1
 			path: src/Sulu/Bundle/AdminBundle/Serializer/Subscriber/TogglerToolbarActionSubscriber.php
 
+		-
+			message: "#^Cannot access offset 'block' on mixed\\.$#"
+			count: 2
+			path: src/Sulu/Bundle/AdminBundle/Tests/Functional/Controller/AdminControllerTest.php
```
Comment (alexander-schranz) [inline phpstan-baseline.neon]: we should avoid any new code to be added to the baseline

**Comment by mamazu** [comment]
Works for me.

File: src/Sulu/Bundle/AdminBundle/Tests/Functional/Controller/AdminControllerTest.php (hunk 1 above)
Comment (mamazu) [reply src/Sulu/Bundle/AdminBundle/Tests/Functional/Controller/AdminControllerTest.php]: I think at this point we might as well put it into a json file and compare results that way. This test is super unreadable. :D

**Review by alexander-schranz — APPROVED** [review]

File: src/Sulu/Bundle/AdminBundle/Tests/Functional/Controller/AdminControllerTest.php [hunk 3]
```diff
@@ -102,7 +104,7 @@ public function testTemplateConfig(): void
 
         // extract json from html
         $applicationElement = $crawler->filter('#application')->first();
-        $this->assertEquals(1, $applicationElement->count(), 'Failed getting the application element');
+        $this->assertCount(2, $applicationElement, 'Failed getting the application element');
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/AdminBundle/Tests/Functional/Controller/AdminControllerTest.php]: This line errors currently 🤔

**Comment by alexander-schranz** [comment]
Thank you all for testing!

**Pull request merged**